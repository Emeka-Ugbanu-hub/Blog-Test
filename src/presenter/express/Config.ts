
import Service from '../../service/Service';
import { LoggerInstance } from 'winston';

export default interface Config {
  readonly morganLogFormat: string,
  readonly morganDirectory: string,
  readonly service: Service,
  readonly logger: LoggerInstance;
}