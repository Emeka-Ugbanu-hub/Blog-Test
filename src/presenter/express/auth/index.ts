import login from './login';
import register from './register';
import forgetPassword from './forgetPassword';

export {
  login, 
  register,
  forgetPassword
}