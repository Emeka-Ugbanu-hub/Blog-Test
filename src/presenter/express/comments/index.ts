import createComment from './createComment';
import getCommentById from './getCommentById';
export {
  createComment,
  getCommentById
}