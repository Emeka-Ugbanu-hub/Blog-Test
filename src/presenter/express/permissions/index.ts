import createPermission from './createPermission';
import getPermissionById from './getPermissionById';
import getPermissions from './getPermissions';
import updatePermission from './updatePermission';
import deletePermissionById from './deletePermissionById';
export {
  getPermissions,
  getPermissionById,
  createPermission,
  updatePermission,
  deletePermissionById
}