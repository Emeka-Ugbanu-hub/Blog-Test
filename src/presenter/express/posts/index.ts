import getPosts from './getPosts';
import createPost from './createPost';
import getPostById from './getPostById';
import updatePost from './updatePost';
import deletePostById from './deletePostById';
export {
  getPosts,
  createPost,
  getPostById,
  updatePost,
  deletePostById
}