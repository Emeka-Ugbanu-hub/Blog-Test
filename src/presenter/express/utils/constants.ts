const OK_200_HTTP_CODE = 200;
const CREATED_201_HTTP_CODE = 201;
const NO_CONTENT_204_HTTP_CODE = 204;
const BAD_REQUEST_400_HTTP_CODE = 400;
const UNAUTHORISED_401_HTTP_CODE = 401;
const FORBIDDEN_403_HTTP_CODE = 403;
const NOT_FOUND_404_HTTP_CODE = 404;
const CONFLICT_409_HTTP_CODE = 409;
const UNPROCESSABLE_ENTITY_422_HTTP_CODE = 422;
const SERVER_ERROR_500_HTTP_CODE = 500;


export {
  OK_200_HTTP_CODE,
  CREATED_201_HTTP_CODE,
  NO_CONTENT_204_HTTP_CODE,
  BAD_REQUEST_400_HTTP_CODE,
  UNAUTHORISED_401_HTTP_CODE,
  FORBIDDEN_403_HTTP_CODE,
  NOT_FOUND_404_HTTP_CODE,
  CONFLICT_409_HTTP_CODE,
  UNPROCESSABLE_ENTITY_422_HTTP_CODE,
  SERVER_ERROR_500_HTTP_CODE,
}