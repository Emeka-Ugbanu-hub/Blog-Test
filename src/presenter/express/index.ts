import { Router, Request, Response} from 'express';
import Config from './Config';
import enhancedRouter from './enhancedRouter';
import {register, login, forgetPassword,resetPassword} from './auth';
import {getPosts, createPost, getPostById, updatePost, deletePostById} from './posts';
import {getUserById, getUsers, updateUser, createUser, deleteUserById} from './users';
import {createComment, getCommentById, getComments, updateComment} from './comments';

export default (config: Config): Router => {
  const router: Router = enhancedRouter(config);

  router.get('/', (req: Request, res: Response) => {
    res.status(200)
       .json({message: "This is where the awesomeness happen..."});
  });

  router.post('/auth/register', register(config));
  router.post('/auth/login', login(config));
  router.post('/auth/forget-password', forgetPassword(config));
  router.post('/auth/reset-password', resetPassword(config));

  router.post('/users', createUser(config));
  router.delete('/users/:user_id', deleteUserById(config));
  router.get('/users/:user_id', getUserById(config));
  router.get('/users', getUsers(config));
  router.patch('/users/:user_id', updateUser(config));
  
  router.get('/posts', getPosts(config));
  router.get('/posts/:post_id', getPostById(config));
  router.delete('/posts/:post_id', deletePostById(config));
  router.post('/posts', createPost(config));
  router.patch('/posts/:post_id', updatePost(config));

  router.get('/comments', getComments(config));
  router.post('/comments', createComment(config));
  router.get('/comments/:comment_id', getCommentById(config));
  router.patch('/comments/:comment_id', updateComment(config));

  return router;
}