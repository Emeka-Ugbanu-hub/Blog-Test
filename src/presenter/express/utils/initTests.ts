import config from '../../../config';
import * as sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
import * as iconvLite from 'iconv-lite';
iconvLite.encodingExists('foo');
import * as express from 'express';
import * as createSupertest from 'supertest';
import logger from '../../../logger';
import presenter from '../index';
import serviceFactory from '../../../service/factory';
import {API_ROUTE_V1} from '../../../utils/constants';

const app: express.Application = express();

const service = serviceFactory();

const presenterFacade = presenter({
  morganLogFormat: config.express.morganLogFormat,
  morganDirectory: config.express.morganDirectory,
  service,
  logger
});

app.use(API_ROUTE_V1, presenterFacade);

const request = createSupertest(app);

export default () => {

  beforeAll(async() => {
    await service.migrate();
  });

  beforeEach(async() => {
    await service.migrate();
  });
  
  afterEach(async() => {
    await service.rollback();
  });

  afterAll(async() => {
    await service.clearService();
  });
  
  return { service, request };
};