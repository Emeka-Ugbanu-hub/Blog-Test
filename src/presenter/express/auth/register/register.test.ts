import initTests from '../../utils/initTests';
import {API_ROUTE_V1, DEFAULT_USER_PERMISSIONS} from '../../../../utils/constants';
import {Response} from 'express';
import {CREATED_201_HTTP_CODE, UNPROCESSABLE_ENTITY_422_HTTP_CODE, CONFLICT_409_HTTP_CODE} from '../../utils/constants';
import config from '../../../../config';
import {TEST_VALID_USER, TEST_INVALID_EMAIL, TEST_VALID_EMAIL, TEST_TOO_SHORT_PASSWORD, TEST_DIFFERENT_VALID_PASSWORD} from '../../../../utils/testValues';
import * as R  from 'ramda';

describe(__filename, () => {

  const { service, request } = initTests();

  it('should fail to create a user without input', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/register`);
    expect(response.status).toBe(UNPROCESSABLE_ENTITY_422_HTTP_CODE);
  });

  it('should fail to create a user when email is invalid', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send({email: TEST_INVALID_EMAIL});
    expect(response.status).toBe(UNPROCESSABLE_ENTITY_422_HTTP_CODE);
  });

  it('should fail to create a user without password', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send({email: TEST_VALID_EMAIL});
    expect(response.status).toBe(UNPROCESSABLE_ENTITY_422_HTTP_CODE);
  });

  it('should fail to create a user with too short password', async () => {
    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send({
                                    email: TEST_VALID_EMAIL, 
                                    password: TEST_TOO_SHORT_PASSWORD
                                  });
    expect(response.status).toBe(UNPROCESSABLE_ENTITY_422_HTTP_CODE);
  });

  it('should fail to create a user with the same email address', async () => {
    const user = await service.register(TEST_VALID_USER);
  
    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send(TEST_VALID_USER);
                                  
    expect(response.status).toBe(CONFLICT_409_HTTP_CODE);
  });

  it('should fail to create a user when the password does not match password_confirmation', async () => {
  
    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send({
                                    email: TEST_VALID_EMAIL, 
                                    password: TEST_VALID_EMAIL,
                                    password_confirmation: TEST_DIFFERENT_VALID_PASSWORD
                                  });
    expect(response.status).toBe(UNPROCESSABLE_ENTITY_422_HTTP_CODE);
  });

  it('should succesfully register a user', async () => {

    const response = await request.post(`${API_ROUTE_V1}/auth/register`)
                                  .send(TEST_VALID_USER);
    const {user, token} = response.body;
    const permissions = await service.getUserPermissions({userId: user.id});
    const permissionsNames = R.pluck('name')(permissions);
    const defaultPermissionsNames =  R.pluck('name')(DEFAULT_USER_PERMISSIONS);
 
    expect(R.intersection(permissionsNames,defaultPermissionsNames).length)
          .toBe(DEFAULT_USER_PERMISSIONS.length);
    expect(response.status).toBe(CREATED_201_HTTP_CODE);
    expect(token).toMatch(/JWT/);
    expect(user.email).toEqual(TEST_VALID_USER.email);
  });

});