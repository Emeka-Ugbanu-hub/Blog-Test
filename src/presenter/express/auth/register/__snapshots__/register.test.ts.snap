// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user when email is invalid 1`] = `
Object {
  "errors": Array [
    "Invalid email in 'user.email'. Received 'undefined'",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user when the password does not match password_confirmation 1`] = `
Object {
  "errors": Array [
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user with the same email address 1`] = `
Object {
  "message": "User already exists",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user with too short password 1`] = `
Object {
  "errors": Array [
    "Required data in user.password must have at least 6 characters.",
    "Missing required value in 'user.password_confirmation'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user without input 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.email'",
    "Missing required value in 'user.password'",
    "Missing required value in 'user.password_confirmation'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;

exports[`/Users/mariuszrajczakowski/Code/rest-api-node-typescript/src/presenter/express/auth/register/register.test.ts should fail to register a user without password 1`] = `
Object {
  "errors": Array [
    "Missing required value in 'user.password'",
    "Missing required value in 'user.password_confirmation'",
    "Passwords must match in user",
  ],
  "message": "Unprocessable entity",
}
`;
